src/msg/async/smc/SMCNetHandler.cc

```diff
@@ -0,0 +1,91 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2014 UnitedStack <haomai@unitedstack.com>
+ *
+ * Author: Haomai Wang <haomaiwang@gmail.com>
+ *
+ * Copyright (C) IBM Corp. 2025
+ *
+ * Author: Aliaksei Makarau <aliaksei.makarau@ibm.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/ip.h>
+#include <netinet/tcp.h>
+#include <arpa/inet.h>
+
+#include "common/debug.h"
+#include "common/errno.h"
+#include "common/dout.h"
+#include "include/compat.h"
+#include "include/sock_compat.h"
+
+#include "SMCNetHandler.h"
+
+#define dout_subsys ceph_subsys_ms
+#undef dout_prefix
+#define dout_prefix *_dout << "SMCNetHandler "
+
+#ifdef HAVE_SMC
+  #ifndef SMCPROTO_SMC
+    #define SMCPROTO_SMC           0       /* SMC protocol, IPv4 */
+    #define SMCPROTO_SMC6          1       /* SMC protocol, IPv6 */
+  #endif
+#endif
+
+namespace ceph {
+namespace smc {
+
+int SMCNetHandler::create_socket(int domain, bool reuse_addr) {
+  int s;
+  int r = 0;
+  int protocol = IPPROTO_TCP;
+
+#ifdef HAVE_SMC
+  /* check if socket is eligible for AF_SMC */
+  if (domain == AF_INET || domain == AF_INET6) {
+    if (domain == AF_INET)
+      protocol = SMCPROTO_SMC;
+    else /* AF_INET6 */
+      protocol = SMCPROTO_SMC6;
+    domain = AF_SMC;
```

Comment:
I wonder if instead of a bunch of cut-and-paste, `PosixStack` and `NetHandler::create_socket()` could just be parameterized.  Just a single boolean -- from a quick look, the only difference between `SMCStack` and `PosixStack` is this 8-line if statement.

Reply:
Thank you!!!!

Reply:
Hi Ilya! @idryomov could it be approached like this: [https://github.com/ibm-s390-cloud/ceph/commit/0f79d4b6d8f41aeb410127bb3c86307e09bdec86](https://github.com/ibm-s390-cloud/ceph/commit/0f79d4b6d8f41aeb410127bb3c86307e09bdec86)  ?

Reply:
Sorry, just realized that I commented on that commit instead of here: https://github.com/ibm-s390-cloud/ceph/commit/0f79d4b6d8f41aeb410127bb3c86307e09bdec86#r166206199